queries/orderbook/barn_batch_rewards.sql

```diff
@@ -96,6 +102,33 @@ order_protocol_fee AS (
                     fp.surplus_factor / (1 - fp.surplus_factor) * surplus -- charge a fraction of surplus
                 )
             END
+            WHEN fp.kind = 'priceimprovement' THEN CASE
+                WHEN os.kind = 'sell' THEN
+                LEAST(
+                    -- at most charge a fraction of volume
+                    fp.price_improvement_max_volume_factor / (1 - fp.price_improvement_max_volume_factor) * os.buy_amount,
+                    -- charge a fraction of price improvement, at most 0
+                    GREATEST(
+                        LEAST(
+                            fp.price_improvement_factor / (1 - fp.price_improvement_factor) * price_improvement,
+                            fp.price_improvement_factor / (1 - fp.price_improvement_factor) * surplus
+                        ),
+                        0
+                    )
+                )
+                WHEN os.kind = 'buy' THEN LEAST(
+                    -- at most charge a fraction of volume
+                    fp.price_improvement_max_volume_factor / (1 + fp.price_improvement_max_volume_factor) * os.sell_amount,
+                    -- charge a fraction of price improvement
+                    GREATEST(
+                        LEAST(
+                            fp.price_improvement_factor / (1 - fp.price_improvement_factor) * price_improvement,
+                            fp.price_improvement_factor / (1 - fp.price_improvement_factor) * surplus
```

Comment:
why is the min needed? Don't we always have `price_improvement <= surplus`?

Reply:
Yes, that should be true. I was not sure how fee policies are set in case of the limit price being better than the quote. But if I understood correctly, in that case we use a `'surplus'` fee policy and never end up in a case where the fee policy is `'priceimprovement'` and `surplus < price_improvement`. So I can get rid of that `least` here. We should add a test for that somewhere, so that we get some error/alert when that implicit assumption is broken in the future.